File: 0075_sort_colors/sort_colors.py
from typing import List


def sort_colors(nums: List[int]):
    """
    Time: O(n)
    Space: O(1)
    """
    c = p0 = 0
    p2 = len(nums) - 1
    while c <= p2:
        if nums[c] == 0:
            nums[c], nums[p0] = nums[p0], nums[c]
            p0 += 1
            c += 1
        elif nums[c] == 1:
            c += 1
        else:
            nums[c], nums[p2] = nums[p2], nums[c]
            p2 -= 1

File: 0075_sort_colors/sort_colors.rb
def sort_colors(nums)
    # Time: O(n)
    # Space: O(1)
    c = c0 = 0
    c2 = nums.length - 1
    while c <= c2
        if nums[c] == 0
            nums[c], nums[c0] = nums[c0], nums[c]
            c0 += 1
            c += 1
        elsif nums[c] == 1
            c += 1
        else
            nums[c], nums[c2] = nums[c2], nums[c]
            c2 -= 1
        end
    end
end